libs/agno/agno/tools/azure_openai.py

```diff
@@ -0,0 +1,179 @@
+
+
+class AzureOpenAITools(Toolkit):
+    """Toolkit for Azure OpenAI services.
+
+    Currently supports:
+    - DALL-E image generation
+    """
+
+    # Define valid parameter options as class constants
+    VALID_MODELS = ["dall-e-3", "dall-e-2"]
+    VALID_SIZES = ["256x256", "512x512", "1024x1024", "1792x1024", "1024x1792"]
+    VALID_QUALITIES = ["standard", "hd"]
+    VALID_STYLES = ["vivid", "natural"]
+
+    def __init__(
+        self,
+        api_key: Optional[str] = None,
+        azure_endpoint: Optional[str] = None,
+        api_version: str = None,
+    ):
+        super().__init__(name="azure_openai")
+
+        # Set credentials from parameters or environment variables
+        self.api_key = api_key or getenv("AZURE_OPENAI_API_KEY")
+        self.azure_endpoint = azure_endpoint or getenv("AZURE_OPENAI_ENDPOINT")
+        self.api_version = api_version or getenv("AZURE_OPENAI_API_VERSION") or "2023-12-01-preview"
+
+        # Log warnings for missing credentials
+        if not self.api_key:
+            logger.error("AZURE_OPENAI_API_KEY not set")
+        if not self.azure_endpoint:
+            logger.error("AZURE_OPENAI_ENDPOINT not set")
+
+        # Register available tools
+        self.register(self.generate_image)
+
+    def _enforce_valid_parameters(self, params: Dict[str, Any]) -> Dict[str, Any]:
+        """Enforce valid parameters by replacing invalid ones with defaults."""
+        enforced = params.copy()
+
+        # Validate model
+        if params.get("model") not in self.VALID_MODELS:
+            enforced["model"] = "dall-e-3"
+            log_debug(f"Enforcing valid model: '{params.get('model')}' -> 'dall-e-3'")
+
+        # Validate size
+        if params.get("size") not in self.VALID_SIZES:
+            enforced["size"] = "1024x1024"
+            log_debug(f"Enforcing valid size: '{params.get('size')}' -> '1024x1024'")
+
+        # Validate quality
+        if params.get("quality") not in self.VALID_QUALITIES:
+            enforced["quality"] = "standard"
+            log_debug(f"Enforcing valid quality: '{params.get('quality')}' -> 'standard'")
+
+        # Validate style
+        if params.get("style") not in self.VALID_STYLES:
+            enforced["style"] = "vivid"
+            log_debug(f"Enforcing valid style: '{params.get('style')}' -> 'vivid'")
+
+        # Validate number of images
+        if not isinstance(params.get("n"), int) or params.get("n", 0) <= 0:
+            enforced["n"] = 1
+            log_debug(f"Enforcing valid n: '{params.get('n')}' -> 1")
+
+        # Special case: dall-e-3 only supports n=1
+        if enforced.get("model") == "dall-e-3" and enforced.get("n", 1) > 1:
+            enforced["n"] = 1
+            log_debug("Enforcing n=1 for dall-e-3 model")
+
+        return enforced
+
+    def generate_image(
+        self,
+        agent: Agent,
+        prompt: str,
+        deployment: Optional[str] = None,
+        model: str = "dall-e-3",
```

Comment:
These should not be function params. Should be on the init of the tool for the user to set, not the model.